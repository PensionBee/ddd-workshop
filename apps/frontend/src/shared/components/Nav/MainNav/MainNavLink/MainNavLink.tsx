import { useRef } from "react";
import { useLocation } from "react-router-dom";
import type { MainNavItemProps } from "../MainNavItem/MainNavItem";
import { StyledItem, StyledLink } from "../MainNavItem/MainNavItem.styled";

export const MainNavLink: React.FC<MainNavItemProps> = ({
  href,
  children,
  $isTabMenu,
  ...props
}) => {
  const linkRef = useRef<HTMLAnchorElement | null>(null);
  const location = useLocation();

  return (
    <StyledItem data-rc="MainNavLink" {...props}>
      <StyledLink
        ref={linkRef}
        href={href}
        $isTabMenu={$isTabMenu}
        $isHighlighted={$isTabMenu && href === location.pathname}
        data-highlighted={$isTabMenu && href === location.pathname}
      >
        {children}
      </StyledLink>
    </StyledItem>
  );
};
