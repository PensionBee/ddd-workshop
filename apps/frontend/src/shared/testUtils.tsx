import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
