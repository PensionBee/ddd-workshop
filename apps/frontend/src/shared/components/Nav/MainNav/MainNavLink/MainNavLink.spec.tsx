import { routerRender } from "@shared/testUtils";
import { MainNavLink } from "./MainNavLink";

const ITEM_CONTENT = "Item Content";
const EXAMPLE_LINK = "/example";

describe("MainNavItem", () => {
  it("should render correctly", () => {
    const { getByText, asFragment } = routerRender(
      <MainNavLink>{ITEM_CONTENT}</MainNavLink>
    );

    expect(getByText(ITEM_CONTENT)).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render with href", () => {
    const { getByText, asFragment } = routerRender(
      <MainNavLink href={EXAMPLE_LINK}>{ITEM_CONTENT}</MainNavLink>
    );

    expect(getByText(ITEM_CONTENT).closest("a")).toHaveAttribute(
      "href",
      EXAMPLE_LINK
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
