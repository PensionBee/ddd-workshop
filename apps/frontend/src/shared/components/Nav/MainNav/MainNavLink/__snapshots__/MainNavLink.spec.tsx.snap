// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MainNavItem > should render correctly 1`] = `
<DocumentFragment>
  <li
    class="sc-jSMdHm iSfdEk"
    data-rc="MainNavLink"
  >
    <a
      class="sc-hKMtDE cQvbKs"
    >
      Item Content
    </a>
  </li>
</DocumentFragment>
`;

exports[`MainNavItem > should render with href 1`] = `
<DocumentFragment>
  <li
    class="sc-jSMdHm iSfdEk"
    data-rc="MainNavLink"
  >
    <a
      class="sc-hKMtDE cQvbKs"
      href="/example"
    >
      Item Content
    </a>
  </li>
</DocumentFragment>
`;

exports[`MainNavItem should render correctly 1`] = `
<DocumentFragment>
  <li
    class="sc-jSUZER dDTVlj"
    data-rc="MainNavLink"
  >
    <a
      class="sc-hLBbgP japqQX"
    >
      Item Content
    </a>
  </li>
</DocumentFragment>
`;

exports[`MainNavItem should render with href 1`] = `
<DocumentFragment>
  <li
    class="sc-jSUZER dDTVlj"
    data-rc="MainNavLink"
  >
    <a
      class="sc-hLBbgP japqQX"
      href="/example"
    >
      Item Content
    </a>
  </li>
</DocumentFragment>
`;
